src/scala/com/github/johnynek/bazel_deps/GradleResolver.scala

```diff
@@ -42,203 +37,214 @@ class GradleResolver(
   // we just want one, so we merge them all
   private def collapseDepTypes(
       lockFile: GradleLockFile
   ): Try[Map[String, GradleLockDependency]] =
-    List(
-      lockFile.compileClasspath,
-      lockFile.annotationProcessor,
-      lockFile.resolutionRules,
-      lockFile.runtimeClasspath,
-      lockFile.testCompileClasspath,
-      lockFile.testRuntimeClasspath
+    TryMerge.tryMergeAll(
+      None,
+      NonEmptyList.of(
+        lockFile.compileClasspath,
+        lockFile.annotationProcessor,
+        lockFile.resolutionRules,
+        lockFile.runtimeClasspath,
+        lockFile.testCompileClasspath,
+        lockFile.testRuntimeClasspath
+      )
     )
     .map(_.getOrElse(Map.empty))
-    .foldM(Map.empty[String, GradleLockDependency]) {
-      TryMerge.tryMerge(None, _, _)
-    }
 
   private def assertConnectedDependencyMap(
       depMap: Map[String, GradleLockDependency]
   ): Try[Unit] =  {
     // Use validated to collect all the missing errors, not just the first
-    type V[+A] = ValidatedNel[(String, String), A]
+    type V[+A] = ValidatedNec[(String, String), A]
 
-    def assertDep(key: String, gld: GradleLockDependency): ValidatedNel[(String, String), Unit] =
+    val unit = Validated.valid(())
+    def assertDep(key: String, gld: GradleLockDependency): ValidatedNec[(String, String), Unit] =
       gld.transitive.getOrElse(Nil).traverse_[V, Unit] { luK =>
         if (!depMap.contains(luK)) {
-          Validated.invalidNel((luK, key))
-        }
-        else {
-          Validated.Valid(())
+          Validated.invalidNec((luK, key))
         }
+        else unit
       }
 
-    // sort to make this deterministically ordered
-    depMap.toList.sortBy(_._1).traverse_[V, Unit] { case (k, v) => assertDep(k, v) } match {
-      case Validated.Valid(()) => Success(())
-      case Validated.Invalid(errs) =>
-        Failure(new Exception("unable to find these referenced transitive deps in dependencies: " +
-          errs.toList.iterator.map { case (luK, key) => s"$key -> $luK" }.mkString("[", ", ", "]")
-        ))
-    }
-}
+    // This looks like depMap.toList.traverse_ but we want to avoid
+    // making a full copy of a giant graph (toList) when in the common case
+    // there are no errors. Instead we collect all the invalid errors
+    // and expand them out
+    depMap.iterator
+      .map { case (k, v) => assertDep(k, v) }
+      .collect { case Validated.Invalid(errsU) => errsU }
+      .toList match {
+        case Nil => Success(())
+        case nonEmpty =>
+          // sort to make this deterministically ordered in the order they appear
+          val errs = nonEmpty.flatMap(_.toList).sortBy(_.swap)
+          Failure(
+            new Exception("unable to find these referenced transitive deps in dependencies: " +
+              errs.iterator.map { case (luK, key) => s"$key -> $luK" }.mkString("[", ", ", "]")
+            )
+          )
+      }
+  }
 
 private def cleanUpMap(
```

Comment:
This method is doing something fairly involved, would be nice to get a scaladoc comment explaining what changes it makes to the map structure, as well as the commented invariant.